core/chains/evm/logpoller/parser.go

```diff
@@ -0,0 +1,528 @@
+
+	where, err := v.whereClause(expressions, limiter)
+	if err != nil {
+		return "", nil, err
+	}
+
+	clauses = append(clauses, where)
+
+	order, err := v.orderClause(limiter)
+	if err != nil {
+		return "", nil, err
+	}
+
+	if len(order) > 0 {
+		clauses = append(clauses, order)
+	}
+
+	limit := v.limitClause(limiter)
+	if len(limit) > 0 {
+		clauses = append(clauses, limit)
+	}
+
+	return strings.Join(clauses, " "), v.args, nil
+}
+
+func (v *pgDSLParser) whereClause(expressions []query.Expression, limiter query.LimitAndSort) (string, error) {
+	segment := "WHERE evm_chain_id = :evm_chain_id"
+
+	if len(expressions) > 0 {
+		exp, err := v.combineExpressions(expressions, query.AND)
+		if err != nil {
+			return "", err
+		}
+
+		segment = fmt.Sprintf("%s AND %s", segment, exp)
+	}
+
+	if limiter.HasCursorLimit() {
+		var op string
+		switch limiter.Limit.CursorDirection {
+		case query.CursorFollowing:
+			op = ">"
+		case query.CursorPrevious:
+			op = "<"
+		default:
+			return "", errors.New("invalid cursor direction")
+		}
+
+		block, txHash, logIdx, err := valuesFromCursor(limiter.Limit.Cursor)
+		if err != nil {
+			return "", err
+		}
+
+		segment = fmt.Sprintf("%s AND block_number %s= :cursor_block AND tx_hash %s= :cursor_txhash AND log_index %s :cursor_log_index", segment, op, op, op)
+
+		v.args.withField("cursor_block_number", block).
+			withField("cursor_txhash", common.HexToHash(txHash)).
+			withField("cursor_log_index", logIdx)
+	}
+
+	return segment, nil
+}
+
+func (v *pgDSLParser) orderClause(limiter query.LimitAndSort) (string, error) {
+	sorting := limiter.SortBy
+
+	if limiter.HasCursorLimit() && !limiter.HasSequenceSort() {
+		var dir query.SortDirection
+
+		switch limiter.Limit.CursorDirection {
+		case query.CursorFollowing:
+			dir = query.Asc
+		case query.CursorPrevious:
+			dir = query.Desc
+		default:
+			return "", errors.New("unexpected cursor direction")
+		}
+
+		sorting = append(sorting, query.NewSortBySequence(dir))
+	}
+
+	if len(sorting) == 0 {
+		return "", nil
+	}
+
+	sort := make([]string, len(sorting))
+
+	for idx, sorted := range sorting {
+		var name string
+
+		order, err := orderToString(sorted.GetDirection())
+		if err != nil {
+			return "", err
+		}
+
+		switch sorted.(type) {
+		case query.SortByBlock:
+			name = blockFieldName
+		case query.SortBySequence:
+			sort[idx] = fmt.Sprintf("block_number %s, tx_hash %s, log_index %s", order, order, order)
+
+			continue
+		case query.SortByTimestamp:
+			name = timestampFieldName
+		default:
+			return "", errors.New("unexpected sort by")
+		}
+
+		sort[idx] = fmt.Sprintf("%s %s", name, order)
+	}
+
+	return fmt.Sprintf("ORDER BY %s", strings.Join(sort, ", ")), nil
+}
+
+func (v *pgDSLParser) limitClause(limiter query.LimitAndSort) string {
+	if !limiter.HasCursorLimit() && limiter.Limit.Count == 0 {
+		return ""
+	}
+
+	return fmt.Sprintf("LIMIT %d", limiter.Limit.Count)
+}
+
+func (v *pgDSLParser) getLastExpression() (string, error) {
```

Comment:
Hmm, I think I have mentioned that in a previous PR, but I believe reusing this instance doesn't give us much value, but can lead to major isues. I'm not an expert when it comes to Go GC, but most of the collectors are blazingly fast and optimized for short-living objects, memory footprint shouldn't be the case here. It would be safer to always create a new `pgDSLParser` on every query parsing. 

For instance, you can have multiple threads (e.g. plugin) running LogPoller read queries, when reusing the same struct it's not gonna work. I believe you should be able to write a concurrent test that spawns multiple routines calling `FilteredLogs` - it should fail

Reply:
Good callout. I'll test concurrency and maybe the best path forward is to benchmark both options.

Reply:
I ran some quick benchmarks on both options with a query containing nested expressions. The benchmark was run with the command `go test -bench=BenchmarkParser -benchtime=10s -benchmem` each time.

```
wordFilter := NewEventByWordFilter(common.HexToHash("0x42"), 8, []primitives.ValueComparator{
	{Value: "a", Operator: primitives.Gt},
	{Value: "b", Operator: primitives.Lte},
})

expressions := []query.Expression{
	{BoolExpression: query.BoolExpression{
		Expressions: []query.Expression{
			query.Timestamp(10, primitives.Eq),
			{BoolExpression: query.BoolExpression{
				Expressions: []query.Expression{
					query.TxHash(common.HexToHash("0x84").Hex()),
					{BoolExpression: query.BoolExpression{
						Expressions: []query.Expression{
							query.Confirmation(primitives.Unconfirmed),
							wordFilter,
						},
						BoolOperator: query.AND,
					}},
				},
				BoolOperator: query.OR,
			}},
		},
		BoolOperator: query.AND,
	}},
}
```

_No changes to `pgDSLParser`. The expression and err are reset for each build of an expression._
goos: darwin
goarch: arm64
pkg: github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller
BenchmarkParser-10       3622035              3270 ns/op            5600 B/op         81 allocs/op
PASS
ok      github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller      20.203s

_Split `pgDSLVistor` from `pgDSLParser` and created a new visitor instance for each expression in the build loop._
goos: darwin
goarch: arm64
pkg: github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller
BenchmarkParser-10       3278098              4672 ns/op            5744 B/op         84 allocs/op
PASS
ok      github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller      38.635s

I don't see a huge difference between the two, but there does appear to still be some time loss on allocations.